File: python/packages/spore_core/src/spore_core/sensor.py
"""SensorChain — post-action feedback controls and output quality evaluation
(issue #10).

Mirrors the Rust reference at ``rust/crates/spore-core/src/sensor.rs``.

Sensors observe the agent's actions (tool calls, tool results, agent
responses) at defined trigger points and emit :class:`SensorResult` values.
The chain is a registry plus a fan-out evaluator: it runs every sensor
registered for a trigger and returns all results without short-circuiting.
The harness decides routing (Warn → inject observation; Halt → stop).

See ``docs/harness-engineering-concepts.md`` § "SensorChain" for the rules
this module enforces.

Rules enforced:

* :meth:`SensorChain.fire` runs every sensor whose ``config.triggers``
  contains the trigger and returns all results — the chain never
  short-circuits.
* :class:`SensorKind.COMPUTATIONAL` sensors run on every matching trigger.
* :class:`SensorKind.INFERENTIAL` sensors are gated by
  ``run_every_n_turns`` (modulo the :attr:`SensorInput.turn_number`) and
  ``run_on_phases`` (if set, the input's ``phase`` must match).
* :meth:`SensorChain.stats` aggregates fire history; ``fire_rate`` is
  ``total_fires / sessions_seen`` clamped to ``[0.0, 1.0]``.
* :meth:`SensorChain.signal_quality_report` flags
  :class:`SensorSignalFlagNeverFired` and
  :class:`SensorSignalFlagAlwaysFiring`.
* Trigger matching for :class:`SensorTriggerPostTool`: an empty
  ``tool_name`` matches any tool; a non-empty ``tool_name`` matches
  exact equality only.
"""

from __future__ import annotations

import asyncio
from collections import defaultdict
from enum import Enum
from typing import Annotated, ClassVar, Literal, NewType, Protocol, runtime_checkable

from pydantic import BaseModel, ConfigDict, Field

from .errors import SporeError
from .harness import SessionId, SessionState
from .memory import Timestamp
from .model import ToolCall, ToolResult
from .tool_registry import TaskPhase

# ============================================================================
# Identity
# ============================================================================

SensorId = NewType("SensorId", str)
"""Stable identifier for a registered sensor."""


def new_sensor_id(s: str) -> SensorId:
    return SensorId(s)


# ============================================================================
# Pydantic base
# ============================================================================


class _Model(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


# ============================================================================
# Enums
# ============================================================================


class SensorKind(str, Enum):
    COMPUTATIONAL = "computational"
    INFERENTIAL = "inferential"


class SensorOutcome(str, Enum):
    PASS = "pass"
    WARN = "warn"
    HALT = "halt"


# ============================================================================
# SensorTrigger (discriminated union on ``kind``)
# ============================================================================


class SensorTriggerPostTool(_Model):
    kind: Literal["post_tool"] = "post_tool"
    tool_name: str = ""


class SensorTriggerPostTurn(_Model):
    kind: Literal["post_turn"] = "post_turn"


class SensorTriggerPostSession(_Model):
    kind: Literal["post_session"] = "post_session"


class SensorTriggerContinuous(_Model):
    kind: Literal["continuous"] = "continuous"


class SensorTriggerOnToolError(_Model):
    kind: Literal["on_tool_error"] = "on_tool_error"


class SensorTriggerOnCompaction(_Model):
    kind: Literal["on_compaction"] = "on_compaction"


SensorTrigger = Annotated[
    SensorTriggerPostTool
    | SensorTriggerPostTurn
    | SensorTriggerPostSession
    | SensorTriggerContinuous
    | SensorTriggerOnToolError
    | SensorTriggerOnCompaction,
    Field(discriminator="kind"),
]


def _trigger_matches(configured: object, fired: object) -> bool:
    """True if a configured trigger matches the trigger that fired.

    ``SensorTriggerPostTool(tool_name="")`` is a wildcard matching any
    ``post_tool`` trigger.
    """
    if isinstance(configured, SensorTriggerPostTool) and isinstance(fired, SensorTriggerPostTool):
        return configured.tool_name == "" or configured.tool_name == fired.tool_name
    return type(configured) is type(fired)


# ============================================================================
# Records
# ============================================================================


class SensorInput(_Model):
    session_id: SessionId
    turn_number: int | None = None
    phase: TaskPhase | None = None
    tool_call: ToolCall | None = None
    tool_result: ToolResult | None = None
    agent_response: str | None = None
    session_state: SessionState = Field(default_factory=SessionState)


class SensorResult(_Model):
    sensor_id: SensorId
    outcome: SensorOutcome
    observation: str | None = None
    detail: str
    fired_at: Timestamp


class SensorSignalThresholds(_Model):
    never_fired_after_n_sessions: int = 10
    always_fired_rate: float = 0.9


class SensorConfig(_Model):
    id: SensorId
    name: str
    kind: SensorKind
    triggers: list[SensorTrigger]
    run_every_n_turns: int | None = None
    run_on_phases: list[TaskPhase] | None = None
    low_signal_threshold: SensorSignalThresholds = Field(default_factory=SensorSignalThresholds)


class SensorStats(_Model):
    sensor_id: SensorId
    total_fires: int
    warn_count: int
    halt_count: int
    pass_count: int
    fire_rate: float
    last_fired: Timestamp | None = None
    low_signal_flag: bool


# ============================================================================
# SensorSignalFlag (discriminated union on ``kind``)
# ============================================================================


class SensorSignalFlagNeverFired(_Model):
    kind: Literal["never_fired"] = "never_fired"
    sensor_id: SensorId
    sessions_observed: int


class SensorSignalFlagAlwaysFiring(_Model):
    kind: Literal["always_firing"] = "always_firing"
    sensor_id: SensorId
    fire_rate: float


SensorSignalFlag = Annotated[
    SensorSignalFlagNeverFired | SensorSignalFlagAlwaysFiring,
    Field(discriminator="kind"),
]


# ============================================================================
# Errors
# ============================================================================


class SensorError(SporeError):
    """Root of every error raised by a :class:`SensorChain`."""

    kind: ClassVar[str] = "SensorError"


class SensorAlreadyRegistered(SensorError):
    kind: ClassVar[str] = "AlreadyRegistered"

    def __init__(self, id: SensorId) -> None:
        self.id = id
        super().__init__(f"sensor already registered: {id!r}")


class SensorValidationFailed(SensorError):
    kind: ClassVar[str] = "ValidationFailed"

    def __init__(self, reason: str) -> None:
        self.reason = reason
        super().__init__(f"validation failed: {reason}")


# ============================================================================
# Protocols
# ============================================================================


@runtime_checkable
class Sensor(Protocol):
    """A single sensor."""

    async def evaluate(self, input: SensorInput) -> SensorResult: ...

    def config(self) -> SensorConfig: ...


@runtime_checkable
class SensorChain(Protocol):
    """Registry + fan-out evaluator."""

    async def register(self, sensor: Sensor) -> None: ...

    async def fire(self, trigger: SensorTrigger, input: SensorInput) -> list[SensorResult]: ...

    async def stats(self, since: Timestamp | None = None) -> list[SensorStats]: ...

    async def signal_quality_report(self, min_sessions: int) -> list[SensorSignalFlag]: ...


# ============================================================================
# Helpers
# ============================================================================


def _inferential_gate_open(cfg: SensorConfig, input: SensorInput) -> bool:
    """Decide whether an inferential sensor should run on this input given
    its gating fields. Computational sensors are unconditional."""
    if cfg.kind == SensorKind.COMPUTATIONAL:
        return True
    if cfg.run_on_phases is not None:
        if len(cfg.run_on_phases) > 0:
            if input.phase is None or input.phase not in cfg.run_on_phases:
                return False
    if cfg.run_every_n_turns is not None:
        n = cfg.run_every_n_turns
        if n == 0:
            return False
        # Fire when turn_number is a multiple of n. Missing turn_number
        # means we cannot gate — default to firing.
        if input.turn_number is not None:
            if input.turn_number % n != 0:
                return False
    return True


# ============================================================================
# StandardSensorChain — reference in-memory implementation
# ============================================================================


class _HistoryRecord:
    __slots__ = ("sensor_id", "session_id", "outcome", "fired_at")

    def __init__(
        self,
        sensor_id: SensorId,
        session_id: SessionId,
        outcome: SensorOutcome,
        fired_at: Timestamp,
    ) -> None:
        self.sensor_id = sensor_id
        self.session_id = session_id
        self.outcome = outcome
        self.fired_at = fired_at


class StandardSensorChain:
    """Reference :class:`SensorChain`. In-memory; suitable for tests and
    short-lived processes."""

    def __init__(self) -> None:
        self._lock = asyncio.Lock()
        self._sensors: list[tuple[SensorConfig, Sensor]] = []
        self._history: list[_HistoryRecord] = []
        self._sessions_seen: set[SessionId] = set()

    # ── register ───────────────────────────────────────────────────────────

    async def register(self, sensor: Sensor) -> None:
        cfg = sensor.config()
        if not cfg.triggers:
            raise SensorValidationFailed("sensor must declare at least one trigger")
        async with self._lock:
            if any(c.id == cfg.id for c, _ in self._sensors):
                raise SensorAlreadyRegistered(cfg.id)
            self._sensors.append((cfg, sensor))

    # ── fire ───────────────────────────────────────────────────────────────

    async def fire(self, trigger: SensorTrigger, input: SensorInput) -> list[SensorResult]:
        async with self._lock:
            self._sessions_seen.add(input.session_id)
            candidates: list[tuple[SensorId, Sensor]] = [
                (cfg.id, sensor)
                for cfg, sensor in self._sensors
                if any(_trigger_matches(t, trigger) for t in cfg.triggers)
                and _inferential_gate_open(cfg, input)
            ]

        results: list[SensorResult] = []
        for sensor_id, sensor in candidates:
            result = await sensor.evaluate(input)
            async with self._lock:
                self._history.append(
                    _HistoryRecord(
                        sensor_id=sensor_id,
                        session_id=input.session_id,
                        outcome=result.outcome,
                        fired_at=result.fired_at,
                    )
                )
            results.append(result)
        return results

    # ── stats ──────────────────────────────────────────────────────────────

    async def stats(self, since: Timestamp | None = None) -> list[SensorStats]:
        async with self._lock:
            sensors = list(self._sensors)
            history = list(self._history)
            sessions_total = len(self._sessions_seen)

        agg_total: dict[SensorId, int] = defaultdict(int)
        agg_pass: dict[SensorId, int] = defaultdict(int)
        agg_warn: dict[SensorId, int] = defaultdict(int)
        agg_halt: dict[SensorId, int] = defaultdict(int)
        agg_last: dict[SensorId, Timestamp] = {}
        for cfg, _ in sensors:
            agg_total[cfg.id] = 0  # ensure presence

        for rec in history:
            if since is not None and str(rec.fired_at) < str(since):
                continue
            agg_total[rec.sensor_id] += 1
            if rec.outcome == SensorOutcome.PASS:
                agg_pass[rec.sensor_id] += 1
            elif rec.outcome == SensorOutcome.WARN:
                agg_warn[rec.sensor_id] += 1
            elif rec.outcome == SensorOutcome.HALT:
                agg_halt[rec.sensor_id] += 1
            agg_last[rec.sensor_id] = rec.fired_at

        out: list[SensorStats] = []
        cfg_by_id = {cfg.id: cfg for cfg, _ in sensors}
        sensor_ids: set[SensorId] = set(agg_total.keys()) | set(cfg_by_id.keys())
        for sid in sensor_ids:
            total = agg_total.get(sid, 0)
            if sessions_total == 0:
                fire_rate = 0.0
            else:
                fire_rate = max(0.0, min(1.0, total / sessions_total))
            cfg = cfg_by_id.get(sid)
            low_signal_flag = False
            if cfg is not None:
                low_signal_flag = fire_rate > cfg.low_signal_threshold.always_fired_rate or (
                    total == 0
                    and sessions_total >= cfg.low_signal_threshold.never_fired_after_n_sessions
                )
            out.append(
                SensorStats(
                    sensor_id=sid,
                    total_fires=total,
                    warn_count=agg_warn.get(sid, 0),
                    halt_count=agg_halt.get(sid, 0),
                    pass_count=agg_pass.get(sid, 0),
                    fire_rate=fire_rate,
                    last_fired=agg_last.get(sid),
                    low_signal_flag=low_signal_flag,
                )
            )
        out.sort(key=lambda s: s.sensor_id)
        return out

    # ── signal_quality_report ──────────────────────────────────────────────

    async def signal_quality_report(self, min_sessions: int) -> list[SensorSignalFlag]:
        async with self._lock:
            sensors = list(self._sensors)
            history = list(self._history)
            sessions_observed = len(self._sessions_seen)

        if sessions_observed < min_sessions:
            return []

        out: list[SensorSignalFlag] = []
        for cfg, _ in sensors:
            fires = [r for r in history if r.sensor_id == cfg.id]
            total = len(fires)
            if sessions_observed == 0:
                fire_rate = 0.0
            else:
                fire_rate = max(0.0, min(1.0, total / sessions_observed))

            if (
                total == 0
                and sessions_observed >= cfg.low_signal_threshold.never_fired_after_n_sessions
            ):
                out.append(
                    SensorSignalFlagNeverFired(
                        sensor_id=cfg.id,
                        sessions_observed=sessions_observed,
                    )
                )
            elif fire_rate > cfg.low_signal_threshold.always_fired_rate:
                out.append(
                    SensorSignalFlagAlwaysFiring(
                        sensor_id=cfg.id,
                        fire_rate=fire_rate,
                    )
                )

        def _sort_key(f: SensorSignalFlag) -> tuple[str, str]:
            if isinstance(f, SensorSignalFlagNeverFired):
                return ("a", str(f.sensor_id))
            return ("b", str(f.sensor_id))

        out.sort(key=_sort_key)
        return out


__all__ = [
    "Sensor",
    "SensorAlreadyRegistered",
    "SensorChain",
    "SensorConfig",
    "SensorError",
    "SensorId",
    "SensorInput",
    "SensorKind",
    "SensorOutcome",
    "SensorResult",
    "SensorSignalFlag",
    "SensorSignalFlagAlwaysFiring",
    "SensorSignalFlagNeverFired",
    "SensorSignalThresholds",
    "SensorStats",
    "SensorTrigger",
    "SensorTriggerContinuous",
    "SensorTriggerOnCompaction",
    "SensorTriggerOnToolError",
    "SensorTriggerPostSession",
    "SensorTriggerPostTool",
    "SensorTriggerPostTurn",
    "SensorValidationFailed",
    "StandardSensorChain",
    "new_sensor_id",
]

File: go/spore-core/sensor/sensor.go
// Package sensor — issue #10 `SensorChain`: post-action feedback controls
// and output quality evaluation.
//
// Sensors observe the agent's actions (tool calls, tool results, agent
// responses) at defined trigger points and emit SensorResults. The chain is
// a registry plus a fan-out evaluator: it runs every sensor registered for
// a trigger and returns all results without short-circuiting. The harness
// decides routing (Warn -> inject observation; Halt -> stop).
//
// See `docs/harness-engineering-concepts.md` § "SensorChain" for the rules
// this package enforces.
//
// Rules enforced
//   - `Fire` runs every sensor whose Config.Triggers contains the trigger
//     and returns all results — the chain never short-circuits.
//   - Computational sensors run on every matching trigger.
//   - Inferential sensors are gated by RunEveryNTurns (modulo
//     SensorInput.TurnNumber) and RunOnPhases (if set, input Phase must
//     match).
//   - `Stats` aggregates fire history; fire_rate = total_fires /
//     sessions_observed clamped to [0.0, 1.0].
//   - `SignalQualityReport` flags:
//   - NeverFired — chain has been asked to fire ≥ min_sessions distinct
//     sessions yet this sensor never fired.
//   - AlwaysFiring — sensor's fire-rate > low_signal_threshold
//     always_fired_rate.
//   - Trigger matching for PostTool: a sensor configured with
//     PostTool{ToolName: ""} (empty) matches any tool. Non-empty ToolName
//     matches exact equality only.
package sensor

import (
	"context"
	"encoding/json"
	"fmt"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/memory"
)

// ============================================================================
// Identity & time
// ============================================================================

// SensorID is the stable identifier for a sensor.
type SensorID string

// Timestamp is re-exported from the memory package for wire compatibility.
type Timestamp = memory.Timestamp

// ============================================================================
// SensorTrigger (tagged via Kind)
// ============================================================================

// SensorTriggerKind discriminates SensorTrigger variants.
type SensorTriggerKind string

const (
	// TriggerKindPostTool — after a specific tool executes.
	TriggerKindPostTool SensorTriggerKind = "post_tool"
	// TriggerKindPostTurn — after every agent turn.
	TriggerKindPostTurn SensorTriggerKind = "post_turn"
	// TriggerKindPostSession — after the loop ends.
	TriggerKindPostSession SensorTriggerKind = "post_session"
	// TriggerKindContinuous — runs outside the change lifecycle.
	TriggerKindContinuous SensorTriggerKind = "continuous"
	// TriggerKindOnToolError — when a tool returns an error.
	TriggerKindOnToolError SensorTriggerKind = "on_tool_error"
	// TriggerKindOnCompaction — after context is compacted.
	TriggerKindOnCompaction SensorTriggerKind = "on_compaction"
)

// SensorTrigger is a tagged union. ToolName is only meaningful for
// TriggerKindPostTool; an empty ToolName matches any tool.
type SensorTrigger struct {
	Kind     SensorTriggerKind `json:"kind"`
	ToolName string            `json:"tool_name,omitempty"`
}

// NewTriggerPostTool returns a PostTool trigger. Empty toolName matches any tool.
func NewTriggerPostTool(toolName string) SensorTrigger {
	return SensorTrigger{Kind: TriggerKindPostTool, ToolName: toolName}
}

// NewTriggerPostTurn returns a PostTurn trigger.
func NewTriggerPostTurn() SensorTrigger { return SensorTrigger{Kind: TriggerKindPostTurn} }

// NewTriggerPostSession returns a PostSession trigger.
func NewTriggerPostSession() SensorTrigger { return SensorTrigger{Kind: TriggerKindPostSession} }

// NewTriggerContinuous returns a Continuous trigger.
func NewTriggerContinuous() SensorTrigger { return SensorTrigger{Kind: TriggerKindContinuous} }

// NewTriggerOnToolError returns an OnToolError trigger.
func NewTriggerOnToolError() SensorTrigger { return SensorTrigger{Kind: TriggerKindOnToolError} }

// NewTriggerOnCompaction returns an OnCompaction trigger.
func NewTriggerOnCompaction() SensorTrigger { return SensorTrigger{Kind: TriggerKindOnCompaction} }

// Matches reports whether this configured trigger matches a fired trigger.
// PostTool{ToolName: ""} is a wildcard matching any PostTool.
func (t SensorTrigger) Matches(fired SensorTrigger) bool {
	if t.Kind != fired.Kind {
		return false
	}
	if t.Kind == TriggerKindPostTool {
		return t.ToolName == "" || t.ToolName == fired.ToolName
	}
	return true
}

// ============================================================================
// SensorKind
// ============================================================================

// SensorKind discriminates the kind of check a sensor performs.
type SensorKind string

const (
	// SensorKindComputational — deterministic, fast (linter, type check, etc).
	SensorKindComputational SensorKind = "computational"
	// SensorKindInferential — probabilistic, slow (LLM-as-judge).
	SensorKindInferential SensorKind = "inferential"
)

// ============================================================================
// SensorOutcome
// ============================================================================

// SensorOutcome is the outcome of a single sensor evaluation.
type SensorOutcome string

const (
	// OutcomePass — no concern.
	OutcomePass SensorOutcome = "pass"
	// OutcomeWarn — continue but inject observation into next turn.
	OutcomeWarn SensorOutcome = "warn"
	// OutcomeHalt — stop execution and surface to human/middleware.
	OutcomeHalt SensorOutcome = "halt"
)

// ============================================================================
// Records
// ============================================================================

// SensorInput is the snapshot a sensor evaluates.
type SensorInput struct {
	SessionID     sporecore.SessionID    `json:"session_id"`
	TurnNumber    *uint32                `json:"turn_number,omitempty"`
	Phase         *sporecore.TaskPhase   `json:"phase,omitempty"`
	ToolCall      *sporecore.ToolCall    `json:"tool_call,omitempty"`
	ToolResult    *sporecore.ToolResult  `json:"tool_result,omitempty"`
	AgentResponse *string                `json:"agent_response,omitempty"`
	SessionState  sporecore.SessionState `json:"session_state"`
}

// NewSensorInput constructs a SensorInput with sensible defaults.
func NewSensorInput(sessionID sporecore.SessionID, state sporecore.SessionState) SensorInput {
	return SensorInput{SessionID: sessionID, SessionState: state}
}

// SensorResult is what a sensor returns.
type SensorResult struct {
	SensorID    SensorID      `json:"sensor_id"`
	Outcome     SensorOutcome `json:"outcome"`
	Observation *string       `json:"observation,omitempty"`
	Detail      string        `json:"detail"`
	FiredAt     Timestamp     `json:"fired_at"`
}

// SensorSignalThresholds tunes the signal-quality report.
type SensorSignalThresholds struct {
	NeverFiredAfterNSessions uint32  `json:"never_fired_after_n_sessions"`
	AlwaysFiredRate          float32 `json:"always_fired_rate"`
}

// DefaultSensorSignalThresholds returns the standard thresholds.
func DefaultSensorSignalThresholds() SensorSignalThresholds {
	return SensorSignalThresholds{
		NeverFiredAfterNSessions: 10,
		AlwaysFiredRate:          0.9,
	}
}

// SensorConfig is the registration shape for a sensor.
type SensorConfig struct {
	ID                 SensorID               `json:"id"`
	Name               string                 `json:"name"`
	Kind               SensorKind             `json:"kind"`
	Triggers           []SensorTrigger        `json:"triggers"`
	RunEveryNTurns     *uint32                `json:"run_every_n_turns,omitempty"`
	RunOnPhases        []sporecore.TaskPhase  `json:"run_on_phases,omitempty"`
	LowSignalThreshold SensorSignalThresholds `json:"low_signal_threshold"`
}

// SensorStats is per-sensor aggregated firing history.
type SensorStats struct {
	SensorID      SensorID   `json:"sensor_id"`
	TotalFires    uint32     `json:"total_fires"`
	WarnCount     uint32     `json:"warn_count"`
	HaltCount     uint32     `json:"halt_count"`
	PassCount     uint32     `json:"pass_count"`
	FireRate      float32    `json:"fire_rate"`
	LastFired     *Timestamp `json:"last_fired,omitempty"`
	LowSignalFlag bool       `json:"low_signal_flag"`
}

// ============================================================================
// SensorSignalFlag (tagged via Kind)
// ============================================================================

// SensorSignalFlagKind discriminates SensorSignalFlag variants.
type SensorSignalFlagKind string

const (
	// FlagKindNeverFired — sensor never fired across observed sessions.
	FlagKindNeverFired SensorSignalFlagKind = "never_fired"
	// FlagKindAlwaysFiring — sensor fires above threshold rate.
	FlagKindAlwaysFiring SensorSignalFlagKind = "always_firing"
)

// SensorSignalFlag is a tagged union emitted by SignalQualityReport.
type SensorSignalFlag struct {
	Kind             SensorSignalFlagKind `json:"kind"`
	SensorID         SensorID             `json:"sensor_id"`
	SessionsObserved uint32               `json:"sessions_observed,omitempty"`
	FireRate         float32              `json:"fire_rate,omitempty"`
}

// NewFlagNeverFired constructs a NeverFired flag.
func NewFlagNeverFired(id SensorID, sessionsObserved uint32) SensorSignalFlag {
	return SensorSignalFlag{Kind: FlagKindNeverFired, SensorID: id, SessionsObserved: sessionsObserved}
}

// NewFlagAlwaysFiring constructs an AlwaysFiring flag.
func NewFlagAlwaysFiring(id SensorID, fireRate float32) SensorSignalFlag {
	return SensorSignalFlag{Kind: FlagKindAlwaysFiring, SensorID: id, FireRate: fireRate}
}

// ============================================================================
// Errors
// ============================================================================

// SensorErrorKind discriminates SensorError variants.
type SensorErrorKind string

const (
	// ErrKindAlreadyRegistered — duplicate sensor id.
	ErrKindAlreadyRegistered SensorErrorKind = "already_registered"
	// ErrKindValidationFailed — invalid sensor config.
	ErrKindValidationFailed SensorErrorKind = "validation_failed"
)

// SensorError is the typed error returned by SensorChain methods.
type SensorError struct {
	Kind     SensorErrorKind `json:"kind"`
	SensorID SensorID        `json:"sensor_id,omitempty"`
	Reason   string          `json:"reason,omitempty"`
}

// Error implements error.
func (e *SensorError) Error() string {
	switch e.Kind {
	case ErrKindAlreadyRegistered:
		return fmt.Sprintf("sensor already registered: %q", string(e.SensorID))
	case ErrKindValidationFailed:
		return fmt.Sprintf("validation failed: %s", e.Reason)
	default:
		return fmt.Sprintf("sensor error: %s", e.Kind)
	}
}

// MarshalJSON emits the Rust-compatible shape: {kind, ...variant fields}.
func (e *SensorError) MarshalJSON() ([]byte, error) {
	switch e.Kind {
	case ErrKindAlreadyRegistered:
		// Rust variant is a tuple struct: AlreadyRegistered(SensorId). Serde
		// serializes this as {"kind":"already_registered","0":SensorId} by
		// default — but we expose it via a named field for Go ergonomics.
		return json.Marshal(struct {
			Kind     SensorErrorKind `json:"kind"`
			SensorID SensorID        `json:"sensor_id"`
		}{e.Kind, e.SensorID})
	case ErrKindValidationFailed:
		return json.Marshal(struct {
			Kind   SensorErrorKind `json:"kind"`
			Reason string          `json:"reason"`
		}{e.Kind, e.Reason})
	}
	return json.Marshal(struct {
		Kind SensorErrorKind `json:"kind"`
	}{e.Kind})
}

// ============================================================================
// Interfaces
// ============================================================================

// Sensor is a single observation unit registered with a SensorChain.
type Sensor interface {
	// Evaluate inspects input and returns a result. Must not mutate input.
	Evaluate(ctx context.Context, input *SensorInput) SensorResult
	// Config returns the sensor's registration config.
	Config() SensorConfig
}

// SensorChain is the registry + fan-out evaluator.
type SensorChain interface {
	// Register validates and inserts a sensor. Duplicate ids return
	// AlreadyRegistered; empty triggers return ValidationFailed.
	Register(ctx context.Context, s Sensor) error

	// Fire runs every sensor whose triggers match and returns every result.
	// Inferential sensors are gated by RunEveryNTurns / RunOnPhases.
	Fire(ctx context.Context, trigger SensorTrigger, input *SensorInput) []SensorResult

	// Stats returns aggregated firing stats. If since is non-nil, only
	// records at-or-after that timestamp are included.
	Stats(ctx context.Context, since *Timestamp) []SensorStats

	// SignalQualityReport flags low-signal sensors. Returns empty when the
	// chain has observed fewer than minSessions sessions.
	SignalQualityReport(ctx context.Context, minSessions uint32) []SensorSignalFlag
}
